### PR TITLE
YarpcMeta service: procedures

### LINKED ISSUE #728: Introspection: inbounds should expose a meta introspection service
This is following the effort of gathering introspection information at the dispatcher. inbounds should expose an introspection service, that let you retrieve some info about exposed services, methods and so on.

JSON seems the most logical encoding to use for this, as its easy to parse for both humans and machines, offers some flexibility, and it doesn't need performance.

Important note: this format should be identical across languages. This is not a full duplicate of the /debug/yarpc page(s), here we only intend to offers information about registered handlers.

### PR DESCRIPTION
**Pull request by bombela**
Exposes procedures registered on a dispatcher. More meta procedures will come in the future.

Closes #728.

### CODE REVIEW

**Comment by bombela** [comment]
Output example (feel free to propose change).

```json
yab -p localhost:28941 -e json meta procedures
{
  "body": {
    "Name": "keyvalue",
    "Procedures": [
      {
        "Encoding": "thrift",
        "Name": "KeyValue::getValue",
        "RPCType": "Unary",
        "Service": "keyvalue",
        "Signature": "GetValue(Key *string) (string)"
      },
      {
        "Encoding": "thrift",
        "Name": "KeyValue::setValue",
        "RPCType": "Unary",
        "Service": "keyvalue",
        "Signature": "SetValue(Key *string, Value *string)"
      },
      {
        "Encoding": "json",
        "Name": "introspect",
        "RPCType": "Unary",
        "Service": "meta",
        "Signature": "introspect() json(dispatcher introspection)"
      },
      {
        "Encoding": "json",
        "Name": "procedures",
        "RPCType": "Unary",
        "Service": "meta",
        "Signature": "procs() json(registered procedures)"
      }
    ],
    "Services": [
      "keyvalue",
      "meta"
    ]
  },
  "ok": true,
  "trace": "0"
}
```
```json
yab -p localhost:28941 -e json meta introspect
{
  "body": {
    "ID": "0xc42007ea00",
    "Inbounds": [
      {
        "Endpoint": "[::]:28941",
        "State": "ChannelListening",
        "Transport": "tchannel"
      },
      {
        "Endpoint": "[::]:24034",
        "State": "Started",
        "Transport": "http"
      }
    ],
    "Name": "keyvalue",
    "Outbounds": [
      {
        "Chooser": {
          "Name": "Single",
          "Peers": [
            {
              "Identifier": "127.0.0.1:24034",
              "State": "Available, 0 pending request(s)"
            }
          ],
          "State": ""
        },
        "Endpoint": "http://127.0.0.1:24034",
        "OutboundKey": "keyvalue1",
        "RPCType": "unary",
        "Service": "keyvalue1",
        "State": "Running",
        "Transport": "http"
      },
      {
        "Chooser": {
          "Name": "",
          "Peers": null,
          "State": ""
        },
        "Endpoint": "localhost:28941",
        "OutboundKey": "keyvalue2",
        "RPCType": "unary",
        "Service": "keyvalue2",
        "State": "Running",
        "Transport": "tchannel"
      }
    ],
    "Procedures": [
      {
        "Encoding": "thrift",
        "HandlerSpec": {},
        "Name": "KeyValue::getValue",
        "Service": "keyvalue",
        "Signature": "GetValue(Key *string) (string)"
      },
      {
        "Encoding": "thrift",
        "HandlerSpec": {},
        "Name": "KeyValue::setValue",
        "Service": "keyvalue",
        "Signature": "SetValue(Key *string, Value *string)"
      },
      {
        "Encoding": "json",
        "HandlerSpec": {},
        "Name": "introspect",
        "Service": "meta",
        "Signature": "introspect() json(dispatcher introspection)"
      },
      {
        "Encoding": "json",
        "HandlerSpec": {},
        "Name": "procedures",
        "Service": "meta",
        "Signature": "procs() json(registered procedures)"
      }
    ]
  },
  "ok": true,
  "trace": "0"
}
```

**Comment by breerly** [comment]
Do you think there is any value in having the procedure name a key in the Services dict?

```
yab -p localhost:28941 -e json meta procedures 
{
  "body": {
    "Name": "keyvalue",
    "Services": {
      "keyvalue": {
        "KeyValue::getValue": {
          "Encoding": "thrift",
          "Name": "KeyValue::getValue",
          "RPCType": "Unary",
          "Service": "keyvalue",
          "Signature": "GetValue(Key *string) (string)"
        },
        "KeyValue::setValue": {
          "Encoding": "thrift",
          "Name": "KeyValue::setValue",
          "RPCType": "Unary",
          "Service": "keyvalue",
          "Signature": "SetValue(Key *string, Value *string)"
        }
...
```

This would allow you to, if you had a procedure in mind, do something like `jq .body.Services.keyvalue.KeyValue::getValue.Signature`, whereas the list doesn't allow for that.

**Comment by bombela** [comment]
You can still do this though:

```jq
jq '.body.Services[][] | select(.Name | contains("getValue"))'
```

**Comment by kriskowal** [comment]
Do you have strong feelings about TwoHump JSON key names? Most other places we use oneHump, including the yab response envelope.

**Comment by bombela** [comment]
I am fine either way, the current output happens to be the default go json one.

**Comment by breerly** [comment]
@bombela fair 👍

**Comment by prashantv** [comment]
I don't have strong opinions on this, but I like `[]procedure` for keeping the output a little simpler. The map does make it easier to interact with `jq`, but since it's still possible with the `select`, and I'm not sure if people will need it that frequently (e.g., people will want to check what names are registered more often I think).

+1 for using `camelCase` with a single hump.

File: yarpcmeta/meta.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		Name      string
+		Handler   interface{}
+		Signature string
+	}{
+		{"procedures", m.procs, `procs() {"Name": "...", "Services": {"...": [{"Name": "..."}]}}`},
+	}
+	var r []transport.Procedure
+	for _, m := range methods {
+		p := json.Procedure(m.Name, m.Handler)[0]
+		p.Service = "yarpc"
```
Comment (breerly) [inline yarpcmeta/meta.go]: FYI we're going to have to reserve the `yarpc` service name, so we should create a validation on the dispatcher that stops anyone from using this name.

**Comment by bombela** [comment]
singleHump

```jq
yab -p http://:24034 -e json yarpc procedures | jq '.body'
{
  "name": "keyvalue",
  "services": {
    "keyvalue": [
      {
        "encoding": "thrift",
        "name": "KeyValue::getValue",
        "rpcType": "Unary",
        "service": "keyvalue",
        "signature": "GetValue(Key *string) (string)"
      },
      {
        "encoding": "thrift",
        "name": "KeyValue::setValue",
        "rpcType": "Unary",
        "service": "keyvalue",
        "signature": "SetValue(Key *string, Value *string)"
      }
    ],
    "yarpc": [
      {
        "encoding": "json",
        "name": "procedures",
        "rpcType": "Unary",
        "service": "yarpc",
        "signature": "procs() {\"name\": \"...\", \"services\": {\"...\": [{\"name\": \"...\"}]}}"
      }
    ]
  }
}
```

File: yarpcmeta/meta.go (hunk 1 above)
Comment (bombela) [inline yarpcmeta/meta.go]: we are not reserving the dispatcher service name, but a sub-service name. to stop anyone using this name would stop us as a side effect...

**Comment by breerly** [comment]
Let's update this PR so that the user still requests the same service name, and we instead namespace the procedure name:

Something like:

```
yab populi meta/procedures
```

Where the string `meta/` here can be: `meta/`, `yarpc/`, `system/`, or `debug/`.

File: dispatcher.go [hunk 2]
```diff
@@ -454,3 +454,13 @@ func (d *Dispatcher) introspect() dispatcherStatus {
...
 	}
 }
+
+// Router returns the procedure router.
+func (d *Dispatcher) Router() transport.Router {
+	return d.table
+}
+
+// Name returns the name of the dispatcher.
+func (d *Dispatcher) Name() string {
```
Comment (breerly) [inline dispatcher.go]: Need a test here.

File: dispatcher.go [hunk 3]
```diff
@@ -454,3 +454,13 @@ func (d *Dispatcher) introspect() dispatcherStatus {
 		Outbounds:  outbounds,
 	}
 }
+
+// Router returns the procedure router.
+func (d *Dispatcher) Router() transport.Router {
```
Comment (breerly) [inline dispatcher.go]: Need a test.

File: yarpcmeta/service.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+// Service exposes dispatcher informations via Procedures().
+type Service struct {
+	disp *yarpc.Dispatcher
+}
+
+// NewMetaService allocates a new yarpc meta service, exposing the list of
+// services and registered procedures of the dispatcher via Procedures().
+func NewMetaService(d *yarpc.Dispatcher) *Service {
```
Comment (breerly) [inline yarpcmeta/service.go]: Let's move this package to `x/` so that we can make breaking changes for awhile before committing.

Comment (breerly) [inline yarpcmeta/service.go]: Also, can't this be private? Looks like just `Register` should be exporte.d

**Comment by bombela** [comment]
Is it safe to assume I should get rid of the "service" in the procedure output? Since we cannot use it for anything in practice anyway.

See https://github.com/yarpc/yarpc-go/issues/742 for some background info.

File: yarpcmeta/service.go (hunk 4 above)
Comment (kriskowal) [inline yarpcmeta/service.go]: Rather than private, consider internal/

File: yarpcmeta/service.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// Procedures returns the procedures to register on a dispatcher.
+func (m *Service) Procedures() []transport.Procedure {
+	methods := []struct {
+		Name      string
+		Handler   interface{}
+		Signature string
+	}{
+		{"yarpc/procedures", m.procs, `procs() {"name": "...", "services": {"...": [{"name": "..."}]}}`},
```
Comment (kriskowal) [inline yarpcmeta/service.go]: My 2¢ is "meta/procedures". Leaves room for "meta/thrift-idl" to complement "Meta::thriftIDL" and "meta/health" to complement "Meta::health", or *vice versa*.

File: yarpcmeta/service.go (hunk 4 above)
Comment (bombela) [inline yarpcmeta/service.go]: Right, so public API is "Register" only, and everything else is internal. "x/yarpcmeta" should be good enough.

File: dispatcher.go (hunk 3 above)
Comment (bombela) [inline dispatcher.go]: tested as a side effect.

File: dispatcher.go (hunk 2 above)
Comment (bombela) [inline dispatcher.go]: tested as a side effect.

**Pull request merged**